copy/copy.go

```diff
@@ -151,65 +170,226 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
+			return errors.Wrap(err, "Error reading signatures")
 		}
-		logrus.Debugf("Source is a manifest list; copying (only) instance %s", instanceDigest)
-		unparsedInstance := image.UnparsedInstance(rawSource, &instanceDigest)
+		sigs = s
+	}
+	if len(sigs) != 0 {
+		c.Printf("Checking if image list destination supports signatures\n")
+		if err := c.dest.SupportsSignatures(); err != nil {
+			return errors.Wrap(err, "Can not copy signatures")
+		}
+	}
 
-		if err := c.copyOneImage(policyContext, options, unparsedInstance); err != nil {
+	// Copy each image, in turn.
+	instanceBlobs := list.Instances()
+	updates := make([]manifest.ListUpdate, len(instanceBlobs))
+	for i, instance := range instanceBlobs {
+		logrus.Debugf("Copying instance %s (%d/%d)", instance.Digest, i+1, len(instanceBlobs))
+		unparsedInstance := image.UnparsedInstance(c.rawSource, &instance.Digest)
+		updatedManifest, updatedManifestType, err := c.copyOneImage(policyContext, options, unparsedInstance, &instance.Digest, &instance.Digest)
+		if err != nil {
+			return err
+		}
+		// Record the result of a possible conversion here.
+		md, err := manifest.Digest(updatedManifest)
+		if err != nil {
 			return err
 		}
+		update := manifest.ListUpdate{
+			Digest:    md,
+			Size:      int64(len(updatedManifest)),
+			MediaType: updatedManifestType,
+		}
+		updates[i] = update
 	}
 
-	if err := c.dest.Commit(); err != nil {
-		return errors.Wrap(err, "Error committing the finished image")
+	// Now apply the updates.
+	if err = list.UpdateInstances(updates); err != nil {
+		return errors.Wrapf(err, "Error updating manifest list")
+	}
+
+	// Check if the updates meaningfully changed the list of images.
+	listIsModified := false
+	if !reflect.DeepEqual(list.Instances(), originalList.Instances()) {
+		listIsModified = true
+	}
+
+	// Determine if we need to convert the manifest to a different format.
+	selectedType, _, err := c.determineListConversion(manifestType, c.dest.SupportedManifestMIMETypes(), "")
+	if err != nil {
+		return errors.Wrapf(err, "Error determining manifest list type to write to destination")
+	}
+	if selectedType != list.MIMEType() {
+		listIsModified = true
+		switch selectedType {
+		case manifest.DockerV2ListMediaType:
+			list, err = list.ToSchema2List()
+			if err != nil {
+				return errors.Wrapf(err, "Error converting manifest list to a schema 2 list")
+			}
+		case imgspecv1.MediaTypeImageIndex:
+			list, err = list.ToOCI1Index()
+			if err != nil {
+				return errors.Wrapf(err, "Error converting manifest list to an oci 1 index")
+			}
+		default:
+			return errors.Errorf("Error converting manifest list to unknown supported type %q", selectedType)
+		}
+	}
+
+	// If we can't use the original value, but we have to change it, flag an error.
+	if listIsModified {
+		canModifyManifest := (len(sigs) == 0)
+		if !canModifyManifest {
+			return errors.Errorf("Internal error: copyMultipleImages() needs to use an updated manifest but that was known to be forbidden")
```

Comment:
The `Internal error` wording in the old code was based on the expectation that `if !canModifyManifest`, the code doing the modification should have already noticed and failed with a more descriptive error message (e.g. refused to convert the manifest). This is not happening here.

Still, the simplest solution may be to drop any signature support for manifest lists, as discussed elsewhere.

Reply:
Still outstanding.

Reply:
Okay, moved this check higher up.